src/edu/tutorials/trainreservation/domain/Train.java

```diff
@@ -85,7 +99,27 @@ private List<Seat> getAvailableSeats(CoachType coachType, LocalDate travelDate)
         return availableSeats;
     }
 
-    public int getTotalDistance() {
-        return destination.getDistance() - source.getDistance();
+    public int calculateDistance(String sourceCity, String destinationCity) {
+        int sourceDistance = getCityDistance(sourceCity);
+        int destinationDistance = getCityDistance(destinationCity);
+        return destinationDistance - sourceDistance;
+    }
+
+    public int getCityDistance(String cityName) {
+        for (City city : cities) {
+            if (city.getName().equals(cityName)) {
+                return city.getDistance();
+            }
+        }
+        return -1;
```

Comment:
distance can never be negative instead it should throw exception. Also, we are duplicating the logic of finding the city instead we should use the getCity() method.